src/core/__tests__/ReactCompositeComponent-test.js

```diff
@@ -599,6 +599,128 @@ describe('ReactCompositeComponent', function() {
 
   });
 
+  it('should warn if context values differ on update', function() {
+    var Parent = React.createClass({
+      childContextTypes: {
+        foo: ReactPropTypes.string
+      },
+
+      getChildContext: function() {
+        return {
+          foo: "bar"
+        };
+      },
+
+      render: function() {
+        return <div>{this.props.children}</div>;
+      }
+    });
+    var Component = React.createClass({
+      contextTypes: {
+        foo: ReactPropTypes.string.isRequired
+      },
+
+      render: function() {
+        return <div />;
+      }
+    });
+
+    var div = document.createElement('div');
+
+    var componentWithSameContext = React.withContext({foo: 'bar'}, function() {
+      return <Component />;
+    });
+    React.render(<Parent>{componentWithSameContext}</Parent>, div);
+
+    expect(console.warn.mock.calls.length).toBe(0);
+
+    var componentWithDifferentContext = React.withContext({foo: 'noise'}, function() {
+      return <Component />;
+    });
+    React.render(<Parent>{componentWithDifferentContext}</Parent>, div);
+
+    expect(console.warn.mock.calls.length).toBe(2);
```

Comment:
Now asserting both, we can think about handling the duplicate warning in another diff.